### PR TITLE
Fix for Issue #1074 - Delta stepping with custom vertices fails

### PR DESCRIPTION
**Pull request by SChudakov**
This commit enables using graphs with custom vertices for DeltaSteppingShortestPath algorithm. Previously this use case failed because ConcurrentSkipListSet required its elements to implement Comparable interface. This is achieved by changing the ConcurrentSkipListSet implementation to a ConcurrentHashSet (a wrapped ConcurrentHashMap, because jkd does not provide the ConcurrentHashSet directly). 

**Performance impact**

There is a benchmark suit DeltaSteppingShortestPathPerformance which compares performance of _DeltaSteppingShortestPath_, _DijkstraShortestPath_ and _BellmanFordShortestPath_, To measure the impact from changing the implementation of the concurrent set it is sufficient to compare just _DeltaSteppingShortestPath_, _DijkstraShortestPath_ algorithms, because _BellmanFordShortestPath_ is nowhere near those two. Below the ratio between the benchmark execution time is provided for all the types of graphs (time(DijkstraShortestPath) / time(DeltaSteppingShortestPath)).

| Graph type     | Graph type parameters               | Before | After | Change % |
|----------------|-------------------------------------|--------|-------|----------|
| BarabasiAlbert | numOfVertices=10000, m=500, m0=1000 | 2,17   | 2,13  | -1       |
| Complete       | numOfVertices=3000                  | 1,65   | 1,84  | +11      |
| Gnm            | numOfVertices=10000, edgeDegree=500 | 2,58   | 2,34  | -9       |
| Gnp            | numOfVertices=10000                 | 2,22   | 2,06  | -7       |
| Stogatz        | numOfVertices=10000, k=1000         | 2,23   | 2,01  | -10      |

As shown in the table the performance degradation is no more than 10%.    

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

**Comment by jsichi** [comment]
In the bug report from @d-michail there was a suggestion to allow a custom comparator to be supplied.  Would it make sense to take a comparator parameter as follows?  If null, then use ConcurrentHashMap.  Otherwise, use ConcurrentSkipListSet with the supplied comparator.

To keep the current default behavior, supply Comparator.naturalOrder().

**Comment by SChudakov** [comment]
Uploaded a new commit. Currently to handle all possible vertex classes the following is done:
- if a Comparator is specified it is used to create ConcurrentSkipListSet
- else if the vertex class implements Comparable, ConcurrentSkipListSet is created
- else the thread-safe hash set is constructed based on the ConcurrentHashMap

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java [hunk 1]
```diff
@@ -570,17 +682,16 @@ private int bucketIndex(double distance)
...
+     * Return the reference to the set that was previously in the bucket.
      *
      * @param bucketIndex bucket index
      * @return content of the bucket
      */
     private Set<V> getContentAndReplace(int bucketIndex)
     {
         Set<V> result = bucketStructure.get(bucketIndex);
-        bucketStructure.set(bucketIndex, new ConcurrentSkipListSet<V>());
+        bucketStructure.set(bucketIndex, Collections.newSetFromMap(new ConcurrentHashMap<>()));
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java:694]: Shouldn't this be using getBucketSupplier?

**Comment by SChudakov** [comment]
Done.

**Comment by jsichi** [comment]
Thanks!

**Pull request merged**